src/os/filestore/FileStore.h

```diff
@@ -381,6 +391,70 @@ class FileStore : public JournalingObjectStore,
     }
   } op_wq;
 
+#ifdef HAVE_LIBAIO
+#define MAX_READ_EVENTS (1024*1024)
+  // See struct aio_info
+  struct read_aio {
+    struct iocb iocb;
+    FDRef fd;
+    bufferlist *bl;
+    bufferptr *bptr;
+    bool done;
+    uint32_t op_flags;
+
+    uint64_t off, len;
+    unsigned alignment;
+    const ghobject_t& oid;
+    Context *context;
+    int status;
+
+    read_aio(FDRef fd_arg, bufferlist *b, uint64_t offset_arg, uint64_t len_arg,
+             unsigned alignment_arg, uint32_t opflags, const ghobject_t& oid_arg,
+             Context *context_arg)
+      : fd(fd_arg), bl(b), done(false),
+        op_flags(opflags), off(offset_arg), len(len_arg), alignment(alignment_arg), oid(oid_arg),
+        context(context_arg), status(0)
+    {
+      // prealloc space for entire read
+      bptr = new bufferptr((unsigned) len + alignment);
+      memset((void*)&iocb, 0, sizeof(iocb));
+    }
+    ~read_aio() {
+      delete bptr;
+    }
+  };
+
+  class ReadCompletionThread : public Thread {
+    FileStore *fs;
+  public:
+    io_context_t read_aio_ctx;
+    atomic_t is_active;
+    ReadCompletionThread(FileStore *fs_arg) :
+      fs(fs_arg), is_active(true) {}
+
+    void * entry() {
+      read_aio_ctx = 0;
+      int r = io_setup(MAX_READ_EVENTS, &read_aio_ctx);
+      if (r < 0) {
+        derr << "ReadCompletionThread: unable to set up io_context for " << this << " "
+	     << cpp_strerror(r) << dendl;
+      } else {
+        fs->read_finish_thread_entry(this);
+        if (is_active.read())
+          derr << "ReadCompletionThread: " << this << "  exit " << dendl;
+      }
+      return 0;
+    };
+  };
+
+  vector<ReadCompletionThread*> read_completion_threads;
```

Comment:
Implementing using std::thread now. However --

I'm having some difficulty implementing read_completion_threads as a vector of objects (not vector of pointers). The main difficulty is that an std::move operator needs to be defined in this case (because creating an object, then doing a push_back() on a vector invokes std::move).
So other fields in the class definition (such as io_context_t, Mutex, etc.) would have to have a std::move operator. At that point, I'm not sure the resulting complications are worth the benefits of avoiding  a vector of pointers.
Does this make sense, or is there an easier way out? Thanks for your suggestion on this.

Also, since ReadCompletionThread adds some extra fields, should it contain std::thread or should it be derived from std::thread? I'm going the former route at the moment.